### PR TITLE
fix: Indiscriminately sending program notifications [DHIS2-17180]

### PR DESCRIPTION
**Pull request by zubaira**
[DHIS2-17180](https://dhis2.atlassian.net/browse/DHIS2-17180)

[DHIS2-17180]: https://dhis2.atlassian.net/browse/DHIS2-17180?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/job/SideEffectTriggerEvent.java [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+ * ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+package org.hisp.dhis.tracker.imports.job;
+
+/**
+ * @author Zubair Asghar
+ */
+public enum SideEffectTriggerEvent {
```
Comment (teleivo) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/job/SideEffectTriggerEvent.java:33]: I assume the `Event` suffix does not refer to a tracker event right? We can avoid that confusion by naming this `SideEffectTrigger` IMHO. Would this still be an accurate name?

**Review by teleivo — CHANGES_REQUESTED** [review]
This should be tested.

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EnrollmentPersister.java [hunk 2]
```diff
@@ -117,6 +118,15 @@ protected boolean isNew(TrackerPreheat preheat, String uid) {
   @Override
   protected TrackerSideEffectDataBundle handleSideEffects(
       TrackerBundle bundle, Enrollment enrollment) {
+
+    SideEffectTrigger trigger = SideEffectTrigger.NONE;
+
+    if (isNew(bundle.getPreheat(), enrollment.getUid())) {
+      trigger = SideEffectTrigger.ENROLLMENT;
+    } else if (enrollment.isCompleted()) {
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EnrollmentPersister.java:126]: This is not checking if the enrollment has been completed now, it is only checking that the status of the current enrollment is `COMPLETED`.
Every time a change will be made to an already completed enrollment, the side effect will be triggered

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EventPersister.java [hunk 3]
```diff
@@ -105,6 +106,15 @@ protected boolean isNew(TrackerPreheat preheat, String uid) {
 
   @Override
   protected TrackerSideEffectDataBundle handleSideEffects(TrackerBundle bundle, Event event) {
+
+    SideEffectTrigger trigger = SideEffectTrigger.NONE;
+
+    if (isNew(bundle.getPreheat(), event.getUid())) {
+      trigger = SideEffectTrigger.EVENT_UPDATE;
+    } else if (event.isCompleted()) {
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EventPersister.java:114]: Same as enrollment

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EventPersister.java [hunk 4]
```diff
@@ -105,6 +106,15 @@ protected boolean isNew(TrackerPreheat preheat, String uid) {
 
   @Override
   protected TrackerSideEffectDataBundle handleSideEffects(TrackerBundle bundle, Event event) {
+
+    SideEffectTrigger trigger = SideEffectTrigger.NONE;
+
+    if (isNew(bundle.getPreheat(), event.getUid())) {
+      trigger = SideEffectTrigger.EVENT_UPDATE;
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EventPersister.java:113]: Why is this called `EVENT_UPDATE`?
This is triggered when a new event is created

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/job/TrackerNotificationThread.java [hunk 5]
```diff
@@ -66,8 +64,11 @@ public TrackerNotificationThread(
     this.manager = manager;
     this.serviceMapper =
         Map.of(
-            Enrollment.class, programNotificationService::sendEnrollmentNotifications,
-            Event.class, programNotificationService::sendEventCompletionNotifications);
+            SideEffectTrigger.ENROLLMENT, programNotificationService::sendEnrollmentNotifications,
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/job/TrackerNotificationThread.java:67]: Do we not react to new/updated events?

Comment (zubaira) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/job/TrackerNotificationThread.java:67]: Only at Enrollment, Enrollment Completion and Event Completion.
Notifications at Event Update can be done using ProgamRule notificaitons.

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EnrollmentPersister.java (hunk 2 above)
Comment (zubaira) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EnrollmentPersister.java]: Yes that is what we discussed in our last call. I am working on it.

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EnrollmentPersister.java [hunk 6]
```diff
@@ -126,6 +142,7 @@ protected TrackerSideEffectDataBundle handleSideEffects(
         .accessedBy(bundle.getUsername())
         .enrollment(enrollment)
         .program(enrollment.getProgram())
+        .triggerEvent(bundle.getPreheat().getTriggers())
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EnrollmentPersister.java:145]: Wouldn't be easier to do everything in `handleSideEffects` method?

You are adding a new abstract method, to add data to the Preheat and then read those data in the `handleSideEffects` method. If you calculate the triggers here, there is no need to add the triggers in the Preheat and to have a new abstract method

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EventPersister.java [hunk 7]
```diff
@@ -114,6 +129,7 @@ protected TrackerSideEffectDataBundle handleSideEffects(TrackerBundle bundle, Ev
         .accessedBy(bundle.getUsername())
         .event(event)
         .program(event.getProgramStage().getProgram())
+        .triggerEvent(bundle.getPreheat().getTriggers())
```
Comment (enricocolasante) [inline dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EventPersister.java:132]: Same as enrollments

**Review by enricocolasante — CHANGES_REQUESTED** [review]

File: dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EnrollmentPersister.java (hunk 6 above)
Comment (zubaira) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EnrollmentPersister.java:145]: yes i did think about it. Problem is `handleSideEffect` is called after `updatePreheat` which makes `isNew` method to returns incorrect result. So either we need to have this separate method or we need to move `handleSideEffect` before `updatePreheat`.

Comment (zubaira) [reply dhis-2/dhis-services/dhis-service-tracker/src/main/java/org/hisp/dhis/tracker/imports/bundle/persister/EnrollmentPersister.java:145]: ok

**Review by enricocolasante — APPROVED** [review]

**Pull request merged**